cpp/src/arrow/util/compression_zlib.cc

```diff
@@ -246,7 +246,7 @@ class GZipCompressor : public Compressor {
     //  again with the same value of the flush parameter and more output space
     //  (updated avail_out), until the flush is complete (deflate returns
     //  with non-zero avail_out)."
-    return FlushResult{bytes_written, (bytes_written == 0)};
+    return FlushResult{bytes_written, (ret == Z_BUF_ERROR && stream_.avail_out == 0)};
```

Comment:
Should be `||`?

Reply:
Hmm, looks like it should just be `avail_out == 0` after all, as I also see this comment in zlib.h:

```
  If deflate returns Z_OK and with
  zero avail_out, it must be called again after making room in the output
  buffer because there might be more output pending
```
Whereas in the comment already cited, there's no mention of the expected return code.

Reply:
Z_BUF_ERROR shouldn't pass silently, though. So either it should return `should_retry = true` or it should return an error.
According to the `deflate()` documentation:
> Note that Z_BUF_ERROR is not fatal, and deflate() can be called again with more input and more output space to continue compressing.

Reply:
Hmm, that makes the new test fail again. Reading that more carefully, it sounds like Z_BUF_ERROR without `avail_out == 0` is just zlib's way to indicate that there was no input to compress, since it states 
> can be called again with **more input** and more output space

Reply:
Ah, if that's the case then I agree we shouldn't signal it.